drivers/ethernet/Kconfig.gd32_hal

```diff
@@ -0,0 +1,155 @@
+# Copyright (c) 2022 Carbon Robotics
+# SPDX-License-Identifier: Apache-2.0
+
+DT_COMPAT_GD_GD32_ETHERNET := gd,gd32-ethernet
+
+menuconfig ETH_GD32_HAL
+	bool "GigaDevice GD32 HAL Ethernet driver"
+	default y if $(dt_compat_enabled,$(DT_COMPAT_GD_GD32_ETHERNET))
+	select USE_GD32_ENET
+	select USE_GD32_SYSCFG
+	help
+	  Enable GigaDevice GD32 HAL based Ethernet driver.
+
+if ETH_GD32_HAL
+
+config ETH_GD32_HAL_RX_THREAD_STACK_SIZE
+	int "RX thread stack size"
+	default 1500
+	help
+	  RX thread stack size
+
+config ETH_GD32_HAL_RX_THREAD_PRIO
+	int "RX thread priority"
+	default 2
+	help
+	  RX thread priority
+
+config ETH_GD32_HAL_PHY_ADDRESS
+	int "PHY address"
+	default 1
+	help
+	  The phy address to use.
+
+choice ETH_GD32_HAL_PHY
+	prompt "PHY chip"
+	default ETH_GD32_HAL_PHY_DP83848
+
+config ETH_GD32_HAL_PHY_DP83848
+	bool "DP83848"
+	help
+	  Use DP83848 PHY.
+
+config ETH_GD32_HAL_PHY_LAN8700
+	bool "LAN8700"
+	help
+	  Use LAN8700 series PHY.
+
+endchoice
```

Comment:
Agreed, why isn't this coming from devicetree (dts/bindings/ethernet/ethernet-phy.yaml)

Reply:
Hi @alsrgv, nice to see you have find the time to submit your work!

For the MDIO part you can take a look at what i have done in https://github.com/sbourdelin/zephyr/commit/183c902203d22f2fb04a974e14828320a2d96ea6

And how i integrate it using the ethernet-phy and DTS in https://github.com/sbourdelin/zephyr/commit/f558a9697bec89f3a7a794ff8bc98457a92a39e2

Note that it was still a WIP tho

Reply:
I see!  That looks nice, I wasn't sure how much progress you've made so far.  I don't think this approach works with their HAL - they're pretty rigid about PHY interaction.

When do you estimate your register version be ready, and would it have PTP support?

Reply:
If GD32 HAL is designed with PHYs hardcoded I hardly see how it can be used. Zephyr drivers should not be constrained by how HALs do things internally. Such HALs may be useful in a bare-metal or freertos, but Zephyr has its own mechanisms, like DT.

In general, I think that low level HALs (like STM32 LL or register based) play well with Zephyr. However, higher level ones tend to be inefficient and force design patterns which are incompatible with Zephyr. Such HALs collide easily with Zephyr because they tend to do things like pin control (e.g., nrfx, infineon), clock control (e.g., TI), power management (e.g., TI, partly nrfx), hardware description, etc. It's nice to have this available when e.g. in FreeRTOS where you have nothing else than a scheduler and a few primitives. However, all these areas have a vendor independent solution in Zephyr: pinctrl, clock_control, PM subsys, Devicetree…

Reply:
It's not entirely true that PHY is hardcoded in this HAL - both PHY address and PHY type can be overridden in the way I'm doing it via Kconfig, and I got it working with 2 PHYs (dev kit and the board we built use different PHYs), it's just not as pretty.  STM32 ethernet driver similarly uses Kconfig to specify PHY address.

What do you guys want to do here?  Abandon this PR?  Or merge it until @sbourdelin PR comes along?